### PR TITLE
Prefer enum Endian instead of arbitrary String in rustc_target::Target

### PR DESCRIPTION
**Pull request by tesuji**
Strong type is better than String I think.

### CODE REVIEW

**Comment by lzutao** [comment]
I'm going to use more enum types for members of Target when it makes sense,
unless people prefer not to.
Discussion (only me currently) could happen in this zulip topic: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/using.20enum.20for.20members.20of.20rustc_target.3A.3Aspec.3A.3ATarget.20.3F

**Comment by petrochenkov** [comment]
Could you move the `enum Endian` to `compiler\rustc_target\src\spec\mod.rs`, or at least reexport it from there, and use imports `use crate::spec::Endian` instead of full paths like `crate::abi::Endian`.

File: compiler/rustc_target/src/abi/mod.rs [hunk 1]
```diff
@@ -234,6 +234,47 @@ pub enum Endian {
...
+        }
+    }
+
+    pub fn is_big(&self) -> bool {
+        Self::Big == *self
+    }
+
+    pub fn is_little(&self) -> bool {
+        Self::Little == *self
+    }
```
Comment (petrochenkov) [inline compiler/rustc_target/src/abi/mod.rs:251]: Could you remove the trivial methods `is_big` and `is_little` and use comparisons like `e == Endian::Little` instead?

File: compiler/rustc_target/src/abi/mod.rs [hunk 2]
```diff
@@ -234,6 +234,47 @@ pub enum Endian {
     Big,
 }
 
+impl Endian {
+    pub fn name(&self) -> &str {
```
Comment (petrochenkov) [inline compiler/rustc_target/src/abi/mod.rs:238]:
```suggestion
    pub fn as_str(&self) -> &str {
```

File: compiler/rustc_target/src/abi/mod.rs [hunk 3]
```diff
@@ -234,6 +234,47 @@ pub enum Endian {
...
+            "little" => Ok(Self::Little),
+            "big" => Ok(Self::Big),
+            _ => Err(format!(r#"unknown endian: "{}""#, s)),
+        }
+    }
+}
+
+impl ToJson for Endian {
+    fn to_json(&self) -> Json {
+        Json::String(self.name().to_owned())
```
Comment (petrochenkov) [inline compiler/rustc_target/src/abi/mod.rs:274]:
```suggestion
        self.as_str().to_json()
```

File: compiler/rustc_codegen_llvm/src/va_arg.rs [hunk 4]
```diff
@@ -52,7 +52,7 @@ fn emit_direct_ptr_va_arg(
     let next = bx.inbounds_gep(addr, &[full_direct_size]);
     bx.store(next, va_list_addr, bx.tcx().data_layout.pointer_align.abi);
 
-    if size.bytes() < slot_size.bytes() && &*bx.tcx().sess.target.target.target_endian == "big" {
+    if size.bytes() < slot_size.bytes() && bx.tcx().sess.target.target.target_endian.is_big() {
```
Comment (pickfire) [inline compiler/rustc_codegen_llvm/src/va_arg.rs:55]: Why not just `*bx.tcx().sess.target.target.target_endian == Endian::Big`? Do we need a helper function to compare value?

Comment (lzutao) [reply compiler/rustc_codegen_llvm/src/va_arg.rs:55]: duplicate https://github.com/rust-lang/rust/pull/77604#discussion_r500578080 .

File: compiler/rustc_target/src/abi/mod.rs (hunk 1 above)
Comment (lzutao) [reply compiler/rustc_target/src/abi/mod.rs:251]: We could, but then we have to import the enum type everywhere
while its only usecase is a member of a struct.
With this boring trivial methods, we don't need import the enum type.

File: compiler/rustc_target/src/abi/mod.rs (hunk 3 above)
Comment (lzutao) [reply compiler/rustc_target/src/abi/mod.rs:274]: It both works the same since I still have to import Json and ToJson
for `impl ToJson for Endian`.

**Comment by lzutao** [comment]
> Could you move the `enum Endian` to `compiler\rustc_target\src\spec\mod.rs`, 

Endian is under `abi` things seems logical to me.

> or at least reexport it from there, 
> and use imports `use crate::spec::Endian` instead of full paths like `crate::abi::Endian`.

There is a MCP to ["Ban `pub` re-exports in compiler"](https://github.com/rust-lang/compiler-team/issues/368),
Also why using `crate::spec::Endian` instead? I don't see the difference. They are all under the same crate.
Or you want to limit access to `abi` module when declaring a target ?

File: compiler/rustc_target/src/abi/mod.rs (hunk 1 above)
Comment (pickfire) [reply compiler/rustc_target/src/abi/mod.rs:251]: I think importing it would probably be better, is it a totally different import?

Comment (lzutao) [reply compiler/rustc_target/src/abi/mod.rs:251]: How? If it's used to assign value to a variable, I would import them. Here we only check the property of it.

**Comment by petrochenkov** [comment]
>There is an MCP to "Ban pub re-exports in compiler",

The first specific application of which (https://github.com/rust-lang/rust/pull/77494) wasn't approved.

>Also why using crate::spec::Endian instead?

That's what specific targets mostly import from, imports will look much cleaner.

(The goal here is to clean things up, but now the PR brings inconsistencies into rustc_target code instead, and does something opposite to a cleanup by introducing the micro-methods.)

**Comment by petrochenkov** [comment]
So I had one more though - perhaps endianness should live in `TargetOptions` instead of `Target`.
The options in `Target` are supposed to lack defaults, but `target_endian` has a very clear default - little endian, so it doesn't need to be explicitly specified by every target.
(That applies to a couple of other options in `Target` as well.)

**Comment by petrochenkov** [comment]
>So I had one more though - perhaps endianness should live in `TargetOptions` instead of `Target`.
>(That applies to a couple of other options in `Target` as well.)

Implemented in https://github.com/rust-lang/rust/pull/77729, I'll mark this PR as waiting on it.

**Comment by lzutao** [comment]
> The first specific application of which (#77494) wasn't approved.

Ok, will do.

> (The goal here is to clean things up, 

Perhaps I didn't make things clear. I've never said this is a clean up. Just that I though everybody prefers strong type when possible. 

> now the PR brings inconsistencies into rustc_target code instead, and does something opposite to a cleanup by introducing the micro-methods.)

So does `Option::is_some`, `is_none`.
The micro methods are used to make things easier for contributors.
If you disagree, I will remove them, just because you're reviewer.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/77729 and https://github.com/rust-lang/rust/pull/78875 have landed, unblocking.

**Pull request closed without merging**